gcloud-java-datastore/src/test/java/com/google/gcloud/datastore/DatastoreTest.java

```diff
@@ -546,7 +550,65 @@ public void testGetArray() {
       // expected - no such property
     }
     assertFalse(result.hasNext());
-    // TODO(ozarov): construct a test to verify more results
+  }
+
+  public void testGetArrayDeferredResults() throws DatastoreRpcException {
+    List<DatastoreV1.Key> keysPb = new ArrayList<>();
+    keysPb.add(KEY1.toPb());
+    keysPb.add(KEY2.toPb());
+    keysPb.add(KEY3.toPb());
+    keysPb.add(KEY4.toPb());
+    keysPb.add(KEY5.toPb());
+    List<DatastoreV1.LookupRequest> lookupRequests = new ArrayList<>();
+    lookupRequests.add(DatastoreV1.LookupRequest.newBuilder().addAllKey(keysPb).build());
+    lookupRequests.add(
+        DatastoreV1.LookupRequest.newBuilder()
+            .addKey(keysPb.get(2))
+            .addKey(keysPb.get(3))
+            .addKey(keysPb.get(5))
+            .build());
+    lookupRequests.add(DatastoreV1.LookupRequest.newBuilder().addKey(keysPb.get(5)).build());
+    Entity entity4 = Entity.builder(KEY4).set("value", StringValue.of("value")).build();
+    Entity entity5 = Entity.builder(KEY5).set("value", "value").build();
+    List<DatastoreV1.LookupResponse> lookupResponses = new ArrayList<>();
+    lookupResponses.add(
+        DatastoreV1.LookupResponse.newBuilder()
+            .addFound(EntityResult.newBuilder().setEntity(ENTITY1.toPb()))
+            .addFound(EntityResult.newBuilder().setEntity(entity4.toPb()))
+            .addDeferred(KEY2.toPb())
+            .addDeferred(KEY3.toPb())
+            .addDeferred(KEY5.toPb())
+            .build());
+    lookupResponses.add(
+        DatastoreV1.LookupResponse.newBuilder()
+            .addFound(EntityResult.newBuilder().setEntity(ENTITY3.toPb()))
+            .addFound(EntityResult.newBuilder().setEntity(entity4.toPb()))
+            .addDeferred(KEY5.toPb())
+            .build());
+    lookupResponses.add(
+        DatastoreV1.LookupResponse.newBuilder()
+            .addFound(EntityResult.newBuilder().setEntity(entity5.toPb()))
+            .build());
+    DatastoreRpcFactory rpcFactoryMock = EasyMock.createStrictMock(DatastoreRpcFactory.class);
+    DatastoreRpc rpcMock = EasyMock.createStrictMock(DatastoreRpc.class);
+    EasyMock.expect(rpcFactoryMock.create(EasyMock.anyObject(DatastoreOptions.class)))
+        .andReturn(rpcMock);
+    for (int i = 0; i < lookupRequests.size(); i++) {
+      EasyMock.expect(rpcMock.lookup(lookupRequests.get(i))).andReturn(lookupResponses.get(i));
+    }
+    EasyMock.replay(rpcFactoryMock, rpcMock);
+    DatastoreOptions options =
+        this.options.toBuilder()
+            .retryParams(RetryParams.getDefaultInstance())
+            .serviceRpcFactory(rpcFactoryMock)
+            .build();
+    Datastore datastore = DatastoreFactory.instance().get(options);
+    Iterator<Entity> iter = datastore.get(KEY1, KEY2, KEY3, KEY4, KEY5);
+    Set<Entity> foundEntities = new HashSet<>();
+    while (iter.hasNext()) {
```

Comment:
key you verify that entities for all keys are returned?

Also, can you add a similar test for fetch (and then also verify that order is maintained even if we get the deferred out of order)?